keps/sig-instrumentation/1748-pod-resource-metrics/README.md

```diff
@@ -0,0 +1,448 @@
+* Visualize batch compute consumption on the cluster where completed pods can be quickly excluded
+
+The implementation would be disablable via a flag in the kube-scheduler `--disable-pod-resource-metrics` to allow distributions to opt-out to the increased cardinality, although in practice we expect the impact in scheduler to be drowned by other O(pods) cardinality in other parts of the system and so this flag should not be necessary.
+
+The metrics would be served by the scheduler that is holding the active scheduler lease. During a failover, no metrics would be reported. This would prevent duplicate metrics from being scraped and aligns with our decision in the scheduler to not fill caches prior to election for resource usage minimization.
+
+Custom schedulers based on the scheduler code should be able to easily reuse and report these metrics, and the configuration of the cluster instrumentation to retrieve those metrics would be left as an exercise for the integrator.
+
+
+### Add recording rules consistent with this metric to describe actual resource usage
+
+In order to make these metrics useful for calculation of actual usage, we would represent the current usage in a form that can easily be queried by applying recording rules in the standard Prometheus stack and equivalents in other systems.
+
+The relevant metric for pod CPU would be:
+
+* `kube_pod_resource_usage_cpu_seconds` - the cpu currently consumed by a running pod since last restart
+  * The value would be in core-seconds of the pod cgroup (including all pods, the pause container, the init container if accounted to the pod, and any other pod accounted CPU the container runtime chooses to assign)
+  * It has the following labels:
+    * `pod` - the name of the pod
+    * `namespace` - the namespace of the pod
+    * `node` - the node the resource is scheduled to, or empty if not yet scheduled
+
+Adding this metric has the secondary benefit of correcting the long tradition of inconsistent calculation of used CPU by people who double count the pod cgroup and the sum of the container cgroups, or who omit the pause container, or omit CPU charged to the pod cgroup.
+
+The relevant metric for pod memory would be:
+
+* `kube_pod_resource_usage_memory_working_set_bytes` - the bytes currently consumed by a running pod
+  * The value would be in bytes accounted to the pod cgroup (including all pods, the pause container, the init container if accounted to the pod, and any other pod accounted memory the container runtime chooses to assign)
+    * Since memory accounting is complex, we would propose working set as the default metric and allow other metrics reported by Kubelet and container runtime to continue to be transformed more granularly
+  * It has the following labels:
+    * `pod` - the name of the pod
+    * `namespace` - the namespace of the pod
+    * `node` - the node the resource is scheduled to, or empty if not yet scheduled
```

Comment:
If this is exposed by a kubelet, we shouldn't expose the node name as a label.  Information about where a metric comes from is generally added by recording rules at scrape time.

Reply:
Yeah but I would expect to define the label, regardless of who is responsible for reporting it.

Reply:
I can clarify the language to say "the following labels SHOULD be reported" or something.